ambry-utils/src/main/java/com.github.ambry.utils/Utils.java

```diff
@@ -721,6 +721,27 @@ public static long addSecondsToEpochTime(long epochTimeInMs, long deltaTimeInSec
     return data;
   }
 
+  /**
+   * Read {@link FileChannel} from a given offset to fill up a {@link ByteBuffer}. Total bytes read should be buffer's
+   * remaining size(limit - position).
+   * @param fileChannel from which data to be read from
+   * @param offset starting offset of the fileChanel to be read from
+   * @param buffer the destination of the read
+   * @throws IOException
+   */
+  public static void readFileToByteBuffer(FileChannel fileChannel, long offset, ByteBuffer buffer) throws IOException {
+    int read = 0;
+    int expectedRead = buffer.remaining();
+    while (buffer.hasRemaining()) {
+      int sizeRead = fileChannel.read(buffer, offset);
+      if (sizeRead == 0 || sizeRead == -1) {
```

Comment:
Is `size == 0` is an exception scenario? (as in, is `FileChannel` necessarily blocking)?

Reply:
I thought fileChannel.read() return 0 only when offset is valid offset and buffer has no space. (This should never happen because we have the check `buffer.hasRemaining()`) if this happen, that's something unexpected, we should throw exception. 

But you are making a good point, if FileChannel is non-blocking, it may return 0. Then I got this: 
https://stackoverflow.com/questions/3955250/why-filechannel-in-java-is-not-non-blocking

Looks like fileChannel is necessarily blocking.

Reply:
Could you log sizeRead as well in the log message?